Turn on mypy for _dynamo/variables/builtin.py

aorenste opened the pull request:
The fact that mypy errors were ignored was hiding several bugs in builtin.py (for example the previous diff's incorrect override and use of `call_getattr`)

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #145554
* #145553
* __->__ #145552
* #145551



cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

aorenste committed:
Turn on mypy for _dynamo/variables/builtin.py

[ghstack-poisoned]

aorenste committed:
Update on "Turn on mypy for _dynamo/variables/builtin.py"

The fact that mypy errors were ignored was hiding several bugs in builtin.py (for example the previous diff's incorrect override and use of `call_getattr`)




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

aorenste commented:
@aorenste has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D68593079).

aorenste committed:
Update on "Turn on mypy for _dynamo/variables/builtin.py"

The fact that mypy errors were ignored was hiding several bugs in builtin.py (for example the previous diff's incorrect override and use of `call_getattr`)




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D68593079](https://our.internmc.facebook.com/intern/diff/D68593079)

[ghstack-poisoned]

XuehaiPan commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -1,4 +1,4 @@
-# mypy: ignore-errors
+# mypy: allow-untyped-defs
 
```
```suggestion

from __future__ import annotations

```

One good practice is to add future annotations to a previously untyped file.

aorenste replied:
That's up for debate. In particular I believe @Skylion007 advises exactly the opposite. The problem (as I understand it) is that the `annotations` import causes all the type annotatios to be treated as strings so it could cause issues with tooling.

Skylion007 replied:
Yeah, this is even more for fun with further typing refinements coming for Python 3.15 and standardizing new typing behavior (which differs from both legacy and future slightly)

XuehaiPan replied:
> The problem (as I understand it) is that the `annotations` import causes all the type annotatios to be treated as strings

We do not rely on the runtime effects of the typing annotations in `torch/_dynamo`. So it is not a problem to use future annotations. This is different from `torch/jit`.

> so it could cause issues with tooling.

No. `mypy` is a static type checker. It does not "run" the code. It does support postponed annotations.

Another benefit to using future annotation is that we can upgrade to the syntax that the future Python supports while keeping a lower Python support. For example, you can use `X | Y` annotation (added in Python 3.10) in Python 3.7+. I see many `change Tuple -> tuple` PRs in the trunk for the Python 3.9 upgrade. If we have future annotations in the first place, we can use the newer annotation syntax and do not need to update the code when we bump the lowest Python version.

Skylion007 replied:
> If we have future annotations in the first place, we can use the newer annotation syntax and do not need to update the code when we bump the lowest Python version.

This depends on the support for annotations in the runtime libraries, once typing_extensions releases functionality like evaluate_forward_ref etc @XuehaiPan then we can safely convert the entire library.

XuehaiPan commented on torch/_dynamo/variables/user_defined.py:
```diff
@@ -98,10 +98,12 @@ def is_forbidden_context_manager(ctx):
 
 
 class UserDefinedClassVariable(UserDefinedVariable):
+    value: type[object]
```
```suggestion
    value: type[Any]
```

aorenste replied:
How is this better?

XuehaiPan replied:
`type[object]` means the `builtins.object` class type, not a generic type that can annotation any class object.

aorenste replied:
This is incorrect.  `type[object]` is the correct type annotation for objects that can take any type.  `type[Any]` is basically the same thing but represents a "poisoned" type such that it allows anything and results in an `Any` type (see the body below of `takes_type_Any`) basically turning off type checking for anything touching it.

```
from typing import Any

class A:
    def __init__(self, i: int) -> None:
        pass
class B: pass

def takes_type_A(x: type[A]) -> None:
    x(5)

def takes_type_obj(x: type[object]) -> None:
    # x(5)  # error: Too many arguments for "object"  [call-arg]
    pass

def takes_type_Any(x: type[Any]) -> None:
    x(5)  # BAD: succeeds without any guarantees of what `x` could be and the resulting object type is `Any`.

# takes_type_A(5)  # error: Argument 1 to "takes_type_A" has incompatible type "int"; expected "type[A]"  [arg-type]
# takes_type_A(int)  # error: Argument 1 to "takes_type_A" has incompatible type "type[int]"; expected "type[A]"  [arg-type]
takes_type_A(A)  # success
# takes_type_A(B)  # error: Argument 1 to "takes_type_A" has incompatible type "type[B]"; expected "type[A]"  [arg-type]

# takes_type_obj(5)  # error: Argument 1 to "takes_type_obj" has incompatible type "int"; expected "type[object]"  [arg-type]
takes_type_obj(int)  # success
takes_type_obj(A)  # success
takes_type_obj(B)  # success

# takes_type_Any(5)  # error: Argument 1 to "takes_type_Any" has incompatible type "int"; expected "type[Any]"  [arg-type]
takes_type_Any(int)  # success
takes_type_Any(A)  # success
takes_type_Any(B)  # success
```

aorenste committed:
Update on "Turn on mypy for _dynamo/variables/builtin.py"

The fact that mypy errors were ignored was hiding several bugs in builtin.py (for example the previous diff's incorrect override and use of `call_getattr`)




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D68593079](https://our.internmc.facebook.com/intern/diff/D68593079)

[ghstack-poisoned]

aorenste committed:
Update on "Turn on mypy for _dynamo/variables/builtin.py"

The fact that mypy errors were ignored was hiding several bugs in builtin.py (for example the previous diff's incorrect override and use of `call_getattr`)




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D68593079](https://our.internmc.facebook.com/intern/diff/D68593079)

[ghstack-poisoned]

aorenste committed:
Update on "Turn on mypy for _dynamo/variables/builtin.py"

The fact that mypy errors were ignored was hiding several bugs in builtin.py (for example the previous diff's incorrect override and use of `call_getattr`)




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D68593079](https://our.internmc.facebook.com/intern/diff/D68593079)

[ghstack-poisoned]

aorenste committed:
Update on "Turn on mypy for _dynamo/variables/builtin.py"

The fact that mypy errors were ignored was hiding several bugs in builtin.py (for example the previous diff's incorrect override and use of `call_getattr`)




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D68593079](https://our.internmc.facebook.com/intern/diff/D68593079)

[ghstack-poisoned]

aorenste committed:
Update on "Turn on mypy for _dynamo/variables/builtin.py"

The fact that mypy errors were ignored was hiding several bugs in builtin.py (for example the previous diff's incorrect override and use of `call_getattr`)




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D68593079](https://our.internmc.facebook.com/intern/diff/D68593079)

[ghstack-poisoned]
